lib/logger/lib/logger.ex

```diff
@@ -813,37 +844,50 @@ defmodule Logger do
   end
 
   defp macro_log(level, data, metadata, caller) do
-    %{module: module, function: fun, file: file, line: line} = caller
+    %{module: module, function: {fun, arity}, file: file, line: line} = caller
 
     caller =
-      compile_time_application_and_file(file) ++
-        [module: module, function: form_fa(fun), line: line]
+      file
+      |> compile_time_application_and_file()
+      |> Keyword.merge(mfa: {module, fun, arity}, line: line)
 
     {compile_metadata, quoted_metadata} =
       if Keyword.keyword?(metadata) do
         metadata = Keyword.merge(caller, metadata)
-        {metadata, metadata}
+        quoted = Keyword.update!(metadata, :mfa, &Macro.escape/1)
+
+        {Map.new(metadata), quote(do: %{unquote_splicing(quoted)})}
       else
-        {[],
+        quoted = Keyword.update!(caller, :mfa, &Macro.escape/1)
+
+        {%{},
          quote do
-           Keyword.merge(unquote(caller), unquote(metadata))
+           Map.merge(%{unquote_splicing(quoted)}, Map.new(unquote(metadata)))
          end}
       end
 
     compile_level = if is_atom(level), do: level, else: :error
 
-    if compile_time_purge_matching?(compile_level, compile_metadata) do
+    legacy = add_legacy_fields(compile_metadata, module, fun, arity)
```

Comment:
Instead of adding them here... could we add them on the handler? We need to extract those in the handler anyway for erlang compat, don't we?

Reply:
No, as this is needed during compilation time for purging. Handler adds them again, but on the runtime.

Reply:
i did a comment on how we can make compilation time purging work in this case. If we want to avoid passing module+function, we can treat them as special keys in the formatter and in the compile time purging.

Reply:
Oops, we are talking about this below. I will resoslve this one.